src/adapter/tstype.ts

```diff
@@ -10,7 +12,7 @@ interface SerialPortOptions {
     baudRate?: number;
     rtscts?: boolean;
     path?: string;
-    adapter?: 'zstack' | 'deconz' | 'zigate';
+    adapter?: 'zstack' | 'deconz' | 'zigate' | typeof Adapter;
```

Comment:
Somehow it feel strange the the adapter is a simple string or an entire javascript class. How do you pass this option from Zigbee2MQTT? Did you also evaluate other solutions?

Reply:
I agree that it's a little strange, but it adds the functionality in without a breaking change. The alternative would be to have this type as just `typeof Adapter` and so the consumer would need pass the adapter class in, but again this is a breaking change.

I'm going to propose that Zigbee2MQTT expose a node API for configuring and starting the service, instead of relying on cloning the repo and an external yaml file (which I find very unidiomatic). I think there is already an issue for this.

Reply:
So how do you currently set it? Wouldn't it be easier to create a temp fork where you add support for this adapter and can be merged later?

Reply:
I personally prefer being able to extend the library is valuable, there may be some adapter that you won't accept a PR for for example. Forks are difficult to manage as well.

Reply:
> Forks are difficult to manage as well especially when this proposed extension to the API works in a backwards compatible way.

I think this can never be assured as the adapter interface can change, so there is always some maintenance here.

I'm still wondering how this can be used from zigbee2mqtt, iobroker.zigbee etc.

Reply:
I'm not familiar with iobroker.zigbee. For zigbee2mqtt, a Node api would be added that would be a separate install path to the current git cloning or docker image. You could then do something to this effect:

```javascript
const zibee2mqtt = require('zibee2mqtt');

zigbee2mqtt.start({
  // Other options here
  adapter: MyCustomAdapter
})
```

Reply:
What if we keep `adapter` as a string and allow an adapter to be loaded from another package?

e.g. `adapter: 'custom/my-own-adapter'`

```
if (serialPortOptions.adapter.startsWith('custom/')) {
  let package = serialPortOptions.adapter.replace('custom/');
  adapter =require(package);
} else {
 .. current logic ...
}
```

Reply:
Sorry for coming in from the side line. I've been looking at the code this weekend, and for me it seems that the tension is created because the class `Controller` implements Controller-logic in addition to interpreting configuration (the properties), creating and injecting dependencies, persisting the database, and, serving as the API for consumers of the library.

Im going to tinker with splitting the internals up while keeping the exposed controller API unchanged. I've joined the discord, so you can reach me there. :)